cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -255,7 +255,7 @@ def _get_one_and_two_qubit_ops_as_timesteps(
                 two_qubit_circuit.append(
                     circuits.Moment() for _ in range(timestep + 1 - len(two_qubit_circuit))
                 )
-                if protocols.num_qubits(op) == 2 and not protocols.is_measurement(op):
+                if protocols.num_qubits(op) >= 2:# and not protocols.is_measurement(op):
```

Comment:
This is incorrect since we could be measuring more than 2 qubits.

Reply:
Thanks for pointing this out! I can change the variable name (two_qubit_... -> multi_qubit_...) and function code to assume multiple qubits are being called and will also see if this assumption was made elsewhere. 

Currently, I am trying to determine a good way to handle intermediate measurements. I've done some research and found that [best_practices.ipynb](https://github.com/quantumlib/Cirq/blob/master/docs/google/best_practices.ipynb) mentions this issue with terminal measurements and gives some advice to either use single qubit measurements or as you mentioned before separating the circuit such that each circuit's final moment is a terminal measurement.

I am going to try and see if I can convert any multi-qubit measurements to single-qubit measurements as this might require the least amount of change and I don't see a reason why single-qubit measurements might be treated differently from multi-qubit measurements on hardware or in the use case. Please let me know if you think this might not work well for this specific case.

Reply:
> I can change the variable name (two_qubit_... -> multi_qubit_...) 

please **don't**

> I can convert any multi-qubit measurements to single-qubit measurements

This will break [ClassicallyControlledOperation](https://quantumai.google/reference/python/cirq/ClassicallyControlledOperation)s

---

My suggestion:
- if the intermediate measurement is a 1 or 2 qubit operation then treat it as any other operation.
- if it measures more than 2 qubits then
   1. break the circuit around it.
   2. recurse on each part and then combine the results.
   3. log a warning message about intermediate measurements.

@tanujkhattar what do you think?

Reply:
here's my current implementation of the suggestion:

```python
@classmethod
    def _get_one_and_two_qubit_ops_as_timesteps(
        cls, circuit: 'cirq.AbstractCircuit'
    ) -> Tuple[List[List['cirq.Operation']], List[List['cirq.Operation']]]:
        """Gets the single and two qubit operations of the circuit factored into timesteps.

        The i'th entry in the nested two-qubit and single-qubit ops correspond to the two-qubit
        gates and single-qubit gates of the i'th timesteps respectively. When constructing the
        output routed circuit, single-qubit operations are inserted before two-qubit operations.
        """
        two_qubit_circuit = circuits.Circuit()

        #TODO: add list variable for intermediate circuits 
        two_qubit_circuits: List[cirq.AbstractCircuit] = []
        prev_two_qubit_moments: int = 0

        single_qubit_ops: List[List[cirq.Operation]] = []
        for moment in circuit:
            for op in moment:

                two_qubit_timestep = two_qubit_circuit.earliest_available_moment(op) 
                single_qubit_timestep = two_qubit_timestep + prev_two_qubit_moments


                single_qubit_ops.extend([] for _ in range(single_qubit_timestep + 1 - len(single_qubit_ops)))
                two_qubit_circuit.append(
                    circuits.Moment() for _ in range(two_qubit_timestep + 1 - len(two_qubit_circuit))
                ) 

                if protocols.num_qubits(op) > 2 and protocols.is_measurement(op):
                    """ TODO: 
                        add intermediate measurment op,  
                        and store circuit,
                        update the timestep length for single qubits correctly,
                        reset single  two_qubit_circuit
                    """
                    two_qubit_circuit[two_qubit_timestep] = two_qubit_circuit[two_qubit_timestep].with_operation(op)
                    two_qubit_circuits.append(two_qubit_circuit)
                    prev_two_qubit_moments += len(two_qubit_circuit.moments)
                    two_qubit_circuit = circuits.Circuit()
                if protocols.num_qubits(op) > 2 and not protocols.is_measurement(op):       
                    two_qubit_circuit[two_qubit_timestep] = two_qubit_circuit[two_qubit_timestep].with_operation(op)
                elif protocols.num_qubits(op) == 2:
                    two_qubit_circuit[two_qubit_timestep] = two_qubit_circuit[two_qubit_timestep].with_operation(op)
                else:
                    single_qubit_ops[single_qubit_timestep].append(op)


        while two_qubit_circuits and len(two_qubit_circuit.moments):
            #TODO: connect end of one circuit to start of another
            prev_circuit = two_qubit_circuits.pop()
            two_qubit_circuit = prev_circuit + two_qubit_circuit

        two_qubit_ops = [list(m) for m in two_qubit_circuit]
        print(two_qubit_ops)
        return two_qubit_ops, single_qubit_ops
```

this results in correct results for measurements on 2 or more qubits but there is still unexpected behavior for single-qubit intermediate measurements on circuits with more than 2 qubits:

Should I try to fix this in this PR as well or open up a separate issue?
```python
import cirq 
import networkx as nx

q0, q1, q2 = cirq.LineQubit.range(3)

circuit = cirq.Circuit(
    cirq.CZ(q0, q1),
    cirq.CZ(q1, q2),
    cirq.CZ(q0, q2),
    cirq.measure(q0, q1, q2, key="key"),
    cirq.H.on_each(q0, q1, q2),
    cirq.measure_each(q0, q1, q2),
)

graph = nx.Graph([[q0, q1], [q1, q2]])
router = cirq.RouteCQC(graph)

print(circuit)
print(router.route_circuit(circuit)[0])
```
prints
```
0: ───@───────@───M('key')───H───M───
      │       │   │
1: ───@───@───┼───M──────────H───M───
          │   │   │
2: ───────@───@───M──────────H───M───
0: ───×───────@───────@───M──────────H───M('q(2)')───
      │       │       │   │
1: ───×───@───@───×───@───M('key')───H───M('q(0)')───
          │       │       │
2: ───────@───────×───────M──────────H───M('q(1)')───
```
```python
import cirq 
import networkx as nx

q0, q1, q2 = cirq.LineQubit.range(3)

circuit = cirq.Circuit(
    cirq.CZ(q0, q1),
    cirq.CZ(q1, q2),
    cirq.CZ(q0, q2),
    cirq.measure_each(q0, q1, q2),
    cirq.H.on_each(q0, q1, q2),
    cirq.measure_each(q0, q1, q2)
)

graph = nx.Graph([[q0, q1], [q1, q2]])
router = cirq.RouteCQC(graph)

print(circuit)
print(router.route_circuit(circuit)[0])
```
prints
```
              ┌──┐
0: ───@────────@─────M───H───M───
      │        │
1: ───@───@────┼M────H───M───────
          │    │
2: ───────@────@─────M───H───M───
              └──┘
0: ───×───────@───────────────────@───M('q(2)')───H───M('q(2)')───
      │       │                   │
1: ───×───@───@───M───H───M───×───@───M('q(0)')───H───M('q(0)')───
          │                   │
2: ───────@───────────────────×───────────────────────────────────
```